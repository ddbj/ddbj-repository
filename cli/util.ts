import { colorize } from 'json_colorize/mod.ts';
import { colors } from 'cliffy/ansi/colors.ts';

export function requireLogin() {
  console.error(`First you need to log in; run ${colors.bold('`ddbj-repository auth login`')}.`);

  Deno.exit(1);
}

export async function ensureSuccess(res: Response) {
  if (res.ok) return;

  const type = res.headers.get('content-type');

  if (type?.includes('application/json')) {
    const payload = await res.json();

    colorize(JSON.stringify(payload, null, 2));
  } else {
    console.error(`Error: ${res.status} ${await res.text()}`);
  }

  Deno.exit(1);
}
