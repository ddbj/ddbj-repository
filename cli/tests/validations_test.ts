import { assertSpyCall, returnsNext, stub } from 'std/testing/mock.ts';
import type { Stub } from 'std/testing/mock.ts';

import { snapshotTest } from 'cliffy/testing/mod.ts';

import mainCommand from '../main_command.ts';
import validationCreateCommand from '../validation_create_command.ts';
import { _internals } from '../util.ts';

import validations1 from './fixtures/validations-1.json' with { type: 'json' };
import validations2 from './fixtures/validations-2.json' with { type: 'json' };

async function runInContext(apiKey: string, responses: Response[], callback: (fetchStub: Stub) => Promise<void>) {
  const originalApiUrl = Deno.env.get('DDBJ_REPOSITORY_API_URL');
  Deno.env.set('DDBJ_REPOSITORY_API_URL', 'http://example.com/api');

  const readStub = stub(_internals, 'read', returnsNext([apiKey]));

  const fetchStub = stub(
    globalThis,
    'fetch',
    returnsNext(responses.map((response) => Promise.resolve(response))),
  );

  try {
    await callback(fetchStub);
  } finally {
    if (originalApiUrl) {
      Deno.env.set('DDBJ_REPOSITORY_API_URL', originalApiUrl);
    }

    readStub.restore();
    fetchStub.restore();
  }
}

await snapshotTest({
  name: 'validation create jvar',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext('dummy', [
      new Response(new File(['foo'], 'foo.txt')), // in createValidation
      new Response(JSON.stringify(validations1[0])), // in createValidation
      new Response(JSON.stringify(validations1[0])), // in waitForRequestFinished
    ], async () => {
      await mainCommand.parse(['validation', 'create', 'jvar', '--excel.file=tests/fixtures/validations-2.json']);
    });
  },
});

await snapshotTest({
  name: 'valiadation list',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext('dummy', [
      new Response(JSON.stringify(validations1), {
        headers: {
          Link: '<http://example.com/api/validations?page=2>; rel="next"',
        },
      }),

      new Response(JSON.stringify(validations2)),
    ], async () => {
      await mainCommand.parse(['validation', 'list']);
    });
  },
});

await snapshotTest({
  name: 'validation show 1',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext('dummy', [
      new Response(JSON.stringify(validations1[0])),
    ], async () => {
      await mainCommand.parse(['validation', 'show', '1']);
    });
  },
});

await snapshotTest({
  name: 'validation show 2',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext('dummy', [
      new Response(JSON.stringify(validations2[0])),
    ], async () => {
      await mainCommand.parse(['validation', 'show', '2']);
    });
  },
});

await snapshotTest({
  name: 'validation get-file 1 dummy-path',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext('dummy', [
      new Response('dummy-file'),
    ], async (fetchStub) => {
      await mainCommand.parse(['validation', 'get-file', '1', 'dummy-path']);

      assertSpyCall(fetchStub, 0, {
        args: ['http://example.com/api/validations/1/files/dummy-path', {
          headers: { 'Authorization': 'Bearer dummy' },
        }],
      });
    });
  },
});
