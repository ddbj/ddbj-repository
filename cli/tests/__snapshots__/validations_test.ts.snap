export const snapshot = {};

snapshot[`valiadation list 1`] = `
stdout:
"ID DB   Created             Started             Finished            Progress Validity Submission
-- --   -------             -------             --------            -------- -------- ----------
1  JVar 2024-01-23 21:43:10 2024-01-23 21:43:11 2024-01-23 21:43:11 finished valid    X-1       
2  JVar 2024-01-23 21:43:10 -                   -                   waiting  -        -         
"
stderr:
""
`;

snapshot[`validation show 1 1`] = `
stdout:
"ID: 1
URL: http://example.com/web/validations/1
DB: JVar
Created: 2024-01-23 21:43:10
Started: 2024-01-23 21:43:11
Finished: 2024-01-23 21:43:11
Progress: finished
Validity: valid
Submission: X-1

Results:
Object File      Validity
_base  -         valid   
Excel  README.md valid   
"
stderr:
""
`;

snapshot[`validation show 2 1`] = `
stdout:
"ID: 2
URL: http://example.com/web/validations/2
DB: JVar
Created: 2024-01-23 21:43:10
Started: -
Finished: -
Progress: waiting
Validity: -
Submission: -

Results:
Object File      Validity
_base  -         -       
Excel  README.md -       
"
stderr:
""
`;

snapshot[`validation get-file 1 dummy-path 1`] = `
stdout:
"dummy-file
"
stderr:
""
`;

snapshot[`validation create jvar 1`] = `
stdout:
'{
  "id": 1,
  "url": "http://localhost:3000/api/validations/1",
  "user": {
    "uid": "keycloak"
  },
  "db": "JVar",
  "created_at": "2024-01-23T12:43:10.750Z",
  "started_at": "2024-01-23T12:43:11.152Z",
  "finished_at": "2024-01-23T12:43:11.272Z",
  "progress": "finished",
  "validity": "valid",
  "objects": [
    {
      "id": "Excel",
      "files": [
        {
          "path": "README.md",
          "url": "http://localhost:3000/api/validations/1/files/README.md"
        }
      ]
    }
  ],
  "results": [
    {
      "object_id": "_base",
      "validity": "valid",
      "details": null,
      "file": null
    },
    {
      "object_id": "Excel",
      "validity": "valid",
      "details": null,
      "file": {
        "path": "README.md",
        "url": "http://localhost:3000/api/validations/1/files/README.md"
      }
    }
  ],
  "submission": {
    "id": "X-1",
    "url": "http://localhost:3000/api/submissions/X-1"
  }
}
'
stderr:
""
`;
